src/osp/Universe.h

```diff
@@ -224,112 +222,62 @@ struct UCompTransformTraj
     std::string m_name;
 };
 
-//struct ACompActivated
-//{
-//    // Index to [probably an active area] when activated
-//    Satellite m_area;
-//}
-
 struct UCompVelocity
 {
     Vector3 m_velocity;
 };
 
 struct UCompType
 {
-    // maybe use an iterator
-    ITypeSatellite* m_type{nullptr};
-};
-
-struct UCompActivatable
-{
-    // put something here some day
-    int m_dummy;
-};
-
-
-/**
- * A specific type of Satellite. A planet, star, vehicle, etc...
- */
-class ITypeSatellite
-{
-public:
-    virtual ~ITypeSatellite() = default;
-    virtual std::string get_name() = 0;
-
-    virtual void add(Satellite sat) = 0;
-    virtual void remove(Satellite sat) = 0;
+    TypeSatIndex m_type{TypeSatIndex::Invalid};
 };
 
-
 /**
- * CRTP to implement common functions of TypeSatellites. add and remove
- * functions will add or remove the component UCOMP_T
+ * Calls rUni.get_reg().emplace<UCOMP_T>(sat), and returns a tuple containing a
+ * reference to the UCOMP_T, or a blank tuple if UCOMP_T is a tag struct.
+ * @tparam UCOMP_T
+ * @param rUni [out] Universe to add new components to
+ * @param sat  [in]  Satellite that is part of rUni
  */
-template<typename TYPESAT_T, typename ... UCOMP_T>
-class CommonTypeSat : public ITypeSatellite
-{
-
-public:
-
-
-    CommonTypeSat(Universe& universe) : m_universe(universe) {}
-    ~CommonTypeSat() = default;
-
-    virtual void add(Satellite sat) { add_get_ucomp(sat); };
-    virtual void remove(Satellite sat) override;
-
-    std::tuple<UCOMP_T& ...> add_get_ucomp_all(Satellite sat);
-    auto& add_get_ucomp(Satellite sat);
-
-    constexpr Universe& get_universe() { return m_universe; }
-
-private:
-
-    Universe& m_universe;
-};
-
-template<typename TYPESAT_T, typename ... UCOMP_T>
-std::tuple<UCOMP_T& ...> CommonTypeSat<TYPESAT_T, UCOMP_T ...>::add_get_ucomp_all(Satellite sat)
+template<typename UCOMP_T>
+constexpr decltype(auto) emplace_filter_tags(Universe& rUni, Satellite sat)
 {
-    auto& type = m_universe.get_reg().get<UCompType>(sat);
-
-    if (type.m_type != nullptr)
+    if constexpr (std::is_empty<UCOMP_T>::value)
     {
-        //return ;
+        // Return empty tuple if tag struct
+        rUni.get_reg().emplace<UCOMP_T>(sat);
+        return std::forward_as_tuple();
     }
-
-    type.m_type = this;
-    return std::forward_as_tuple((m_universe.get_reg().emplace<UCOMP_T>(sat)) ...);
-}
-
-template<typename TYPESAT_T, typename ... UCOMP_T>
-auto& CommonTypeSat<TYPESAT_T, UCOMP_T ...>::add_get_ucomp(Satellite sat)
-{
-    auto& type = m_universe.get_reg().get<UCompType>(sat);
-
-    if (type.m_type != nullptr)
+    else
     {
-        //return ;
+        // Return reference tuple for normal struct
+        UCOMP_T &comp = rUni.get_reg().emplace<UCOMP_T>(sat);
+        return std::forward_as_tuple(comp);
     }
-
-    type.m_type = this;
-    return std::get<0>(std::forward_as_tuple((m_universe.get_reg().emplace<UCOMP_T>(sat)) ...));
 }
 
 template<typename TYPESAT_T, typename ... UCOMP_T>
```

Comment:
I'm having a hard time with this naming convention for UCOMP_T

I feel like it should be plural, but what's the "U" prefix represent? It's used in multiple places.

Reply:
"UComp" is for "Universe Component" structs for use in ECS. It parallels "AComp" for "Active Component."

Reply:
Good to know.

But in this case, UCOMP_T represents several UCOMP_T template parameters, which is why i recommended UCOMPS_T